martin/src/cog/source.rs

```diff
@@ -456,6 +480,79 @@ fn get_origin(
     }
 }
 
+fn get_full_resolution(
+    pixel_scale: Option<&[f64]>,
+    transformation: Option<&[f64]>,
+    path: &Path,
+) -> Result<[f64; 2], CogError> {
+    match (pixel_scale, transformation) {
+        // ModelPixelScaleTag = (ScaleX, ScaleY, ScaleZ)
+        (Some(scale), _) => Ok([scale[0], scale[1]]),
+        (_, Some(matrix)) => {
+            let mut x_res =
+                (matrix[0] * matrix[0] + matrix[4] * matrix[4] + matrix[8] * matrix[8]).sqrt();
+            x_res = x_res.copysign(matrix[0]);
+            let mut y_res =
+                (matrix[1] * matrix[1] + matrix[5] * matrix[5] + matrix[9] * matrix[9]).sqrt();
+            // A positive y_res indicates that model space Y cordinates decrease as raster space J indices increase. This is the standard vertical relationship between raster space and model space
+            y_res = y_res.copysign(-matrix[5]);
+            Ok([x_res, y_res]) // drop the z scale directly as we don't use it
+        }
+        (None, None) => Err(CogError::GetFullResolutionFailed(path.to_path_buf())),
+    }
+}
+
+fn raster2model(i: u32, j: u32, matrix: &[f64]) -> (f64, f64) {
+    let i = f64::from(i);
+    let j = f64::from(j);
+    let x = matrix[3] + (matrix[0] * i) + (matrix[1] * j);
+    let y = matrix[7] + (matrix[4] * i) + (matrix[5] * j);
+    (x, y)
+}
+
+fn get_extent(
+    origin: &[f64; 3],
+    transformation: Option<&[f64]>,
+    (full_width_pixel, full_height_pixel): (u32, u32),
+    (full_width, full_height): (f64, f64),
+) -> [f64; 4] {
+    if let Some(matrix) = transformation {
+        let corners = [
+            [0, 0],
+            [0, full_height_pixel],
+            [full_width_pixel, 0],
+            [full_width_pixel, full_height_pixel],
+        ];
+        let transformed = corners.map(|pixel| raster2model(pixel[0], pixel[1], matrix));
+        let max_x = transformed
+            .iter()
+            .map(|(x, _)| *x)
+            .max_by(f64::total_cmp)
+            .expect("corners has >1 element => max exists");
+        let min_x = transformed
+            .iter()
+            .map(|(x, _)| *x)
+            .min_by(f64::total_cmp)
+            .expect("corners has >1 element => min exists");
+        let max_y = transformed
+            .iter()
+            .map(|(_, y)| *y)
+            .max_by(f64::total_cmp)
+            .expect("corners has >1 element => max exists");
+        let min_y = transformed
+            .iter()
+            .map(|(_, y)| *y)
+            .min_by(f64::total_cmp)
+            .expect("corners has >1 element => min exists");
```

Comment:
@sharkAndshark what do you think about this instead?

Reply:
thx! I've made small updates based on your suggestion